lib/livebook/app.ex

```diff
@@ -45,28 +51,56 @@ defmodule Livebook.App do
           started_by: Livebook.Users.User.t() | nil
         }
 
-  use GenServer, restart: :temporary
+  @typedoc """
+  Notebook and related information for deploying an app version.
+
+  This information is used to start app sessions.
+
+    * `:notebook` - the notebook to use for the deployment
+
+    * `:files_tmp_path` - a path to local directory with notebook files.
+      This should be a temporary copy of the files that the app process
+      takes ownership over. The app is going to remove the directory
+      once no longer needed.
+
+      The app uses this local copy of the files, so that changes to
+      the original files can be made safely. Also, if the original
+      files are stored on a remote file system, we want to download
+      them once, rather than in each starting session. Finally, some
+      app specs (such as Teams) may need to unpack their files into
+      a temporary directory and only the app process knows when to
+      remove this directory
+
+    * `:app_spec` - the app spec that was used to load the notebook
+
+    * `:permanent` - whether the app is being deployed as a permanent,
+      typically by `Livebook.Apps.Manager`
+
+    * `:warnings` - a list of warnings to show for the deployment
+
+  """
+  @type deployment_bundle :: %{
+          notebook: Livebook.Notebook.t(),
+          files_tmp_path: String.t(),
+          app_spec: Livebook.Apps.AppSpec.t(),
+          permanent: boolean(),
+          warnings: list(String.t())
+        }
 
   @doc """
   Starts an apps process.
 
   ## Options
 
-    * `:notebook` (required) - the notebook for initial deployment
-
-    * `:warnings` - a list of warnings to show for the initial deployment
-
-    * `:files_source` - a location to fetch notebook files from, see
-      `Livebook.Session.start_link/1` for more details
+    * `:deployment_bundle` (required) - see `t:deployment_bundle/0`
 
   """
   @spec start_link(keyword()) :: {:ok, pid} | {:error, any()}
   def start_link(opts) do
-    notebook = Keyword.fetch!(opts, :notebook)
-    warnings = Keyword.get(opts, :warnings, [])
-    files_source = Keyword.get(opts, :files_source)
+    opts = Keyword.validate!(opts, [:deployment_bundle])
+    deployment_bundle = Keyword.fetch!(opts, :deployment_bundle)
 
-    GenServer.start_link(__MODULE__, {notebook, warnings, files_source})
+    GenServer.start_link(__MODULE__, {deployment_bundle})
```

Comment:
Nit: you don't need a tuple:

```suggestion
    GenServer.start_link(__MODULE__, deployment_bundle)
```